text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+
+This RFC proposes to loosen this restriction to allow custom smart pointer types to be accepted as a `self` type just like for the standard library types.
+
+See also [this blog post](https://medium.com/@adetaylor/the-case-for-stabilizing-arbitrary-self-types-b07bab22bb45), especially for a list of more specific use-cases.
+
+## Motivation for the v2 changes
+
+Unstable Rust contains an implementation of arbitrary self types based around the `Deref<Target=T>` trait. Naturally, that trait also provides a means to create a `&T`.
+
+However, if it's OK to create a reference `&T`, you _probably don't need this feature_. You can probably simply use `&self` as your receiver type.
+
+This feature is fundamentally aimed at smart pointer types `P<T>` where it's not safe to create a reference `&T`. As noted above, that's most commonly because of semantic differences to pointers in other languages, but it might be because references have special meaning or behavior in some pure Rust domain. Either way, it's not OK to create a Rust reference `&T` or `&mut T`, yet we may want to allow methods to be called on some reference-like thing.
+
+For this reason, implementing `Deref::deref` is problematic for _nearly everyone who wants to use arbitrary self types_.
+
+If you're implementing a smart pointer `P<T>` yet you can't allow a reference `&T` to exist, any option for implementing `Deref::deref` has drawbacks:
+
+* Specify `Deref::Target=T` and panic in `Deref::deref`. Not good.
+* Specify `Deref::Target=*const T`. This works with the current arbitrary self types feature, but is only possible if your smart pointer type contains a `*const T` which you can reference - this isn't the case for (for instance) weak pointers or types containing `NonNull`.
+
+Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring a method, users can declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>` or `Self`.
+
+The `Receiver` trait is simple and only requires to specify the `Target` type:
+
+```rust
+trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+The `Receiver` trait is already implemented for many standard library types:
+- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<Ptr<Self>>` (and in fact, any type which implements `Deref`)
+- references: `&Self` and `&mut Self`
+- pointers: `*const Self` and `*mut Self`
+
+Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.
+
+All of the following self types are valid:
+
+```rust
+impl Foo {
+    fn by_value(self /* self: Self */);
+    fn by_ref(&self /* self: &Self */);
+    fn by_ref_mut(&mut self /* self: &mut Self */);
+    fn by_ptr(*const Self);
```

Comment:
Is this meant to be `self: *const Self`, or a new shorthand for pointer self params `*const self`?

Reply:
Fixed, thanks!